### PR TITLE
Allocation subsampling algorithm

### PR DESCRIPTION
**Pull request by johnbley**
Introduce an algorithm for subsampling of allocations, with unit tests.  Algorithm described in more detail in comments, but note that it isn't wired up to anything yet since we should have a discussion about the constants involved and/or configuration of them.

### CODE REVIEW

**Review by pjanotti — APPROVED** [review]
LGTM. One thing in my mind is that we may have a lot of contention on that lock when using server GC. That said seems reasonable to start with something correct and simple and optimize as needed.

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 1]
```diff
@@ -752,6 +752,57 @@ void CaptureAllocationStack(AlwaysOnProfiler* prof, ThreadSamplesBuffer* buffer)
     }
 }
 
+AllocationSubSampler::AllocationSubSampler(uint32_t targetPC, uint32_t secondsPC) :
```
Comment (Kielek) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:755]: Different parameter names than in header file.

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 2]
```diff
@@ -752,6 +752,57 @@ void CaptureAllocationStack(AlwaysOnProfiler* prof, ThreadSamplesBuffer* buffer)
...
+    targetPerCycle(targetPC), secondsPerCycle(secondsPC), seenThisCycle(0), sampledThisCycle(0), seenLastCycle(0),
+    nextCycleStartMillis(
+        std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch())),
+    sampleLock(), rand()
```
Comment (Kielek) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:759]: Rand is always seed by default value - 1 - https://cplusplus.com/reference/random/linear_congruential_engine/linear_congruential_engine/
Is it expected?

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 3]
```diff
@@ -752,6 +752,57 @@ void CaptureAllocationStack(AlwaysOnProfiler* prof, ThreadSamplesBuffer* buffer)
...
+}
+void AllocationSubSampler::AdvanceCycle(std::chrono::milliseconds now)
+{
+    nextCycleStartMillis = now + std::chrono::seconds(secondsPerCycle);
+    seenLastCycle = seenThisCycle;
+    seenThisCycle = 0;
+    sampledThisCycle = 0;
+}
+
+// We want to sample T items out of N per unit time, where N is unknown and may be < S or may be orders
```
Comment (Kielek) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:770]: `S` is not defined here, same in the next line.

**Review by Kielek — APPROVED** [review]

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp (hunk 2 above)
Comment (johnbley) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp]: Whoops!  Assumed the default would be more sane than that; fixed to seed with time.

**Review by Kielek — APPROVED** [review]

**Comment by johnbley** [comment]
> LGTM. One thing in my mind is that we may have a lot of contention on that lock when using server GC. That said seems reasonable to start with something correct and simple and optimize as needed.

Agreed, though I think in that scenario (heavily contended AllocationTicks) I think the buffer lock might be a bigger issue.

**Pull request merged**